datastore/openidmanager.go

```diff
@@ -0,0 +1,102 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package datastore
+
+import (
+	"errors"
+	"log"
+	"time"
+)
+
+// OpenidNonceMaxAge is the maximum age of stored nonces. Any nonces older
+// than this will automatically be rejected. Stored nonces older
+// than this will periodically be purged from the database.
+const OpenidNonceMaxAge = maxNonceAgeInSeconds * time.Second
+const maxNonceAgeInSeconds = 60
+
+const createOpenidNonceTableSQL = `
+	CREATE TABLE IF NOT EXISTS openidnonce (
+		id             serial primary key not null,
+		nonce          varchar(255) not null,
+		endpoint       varchar(255) not null,
+		timestamp      int not null
+	)
+`
+
+// Indexes
+const createOpenidNonceIndexSQL = "CREATE UNIQUE INDEX IF NOT EXISTS nonce_endpoint_idx ON openidnonce (nonce, endpoint)"
+
+// Queries
+const createOpenidNonceSQL = "INSERT INTO openidnonce (nonce, endpoint, timestamp) VALUES ($1, $2, $3)"
+const deleteExpiredOpenidNonceSQL = "DELETE FROM openidnonce where timestamp<$1"
+
+// OpenidNonce holds the details of the nonce, combining a timestamp and random text
```

Comment:
There are checks done to stop replay attacks and that combines the nonce string and the endpoint. I'll add a link to the docs... not that they are very clear!